File: web-designer/tcmenugen/src/domain/PersistedMenu.ts
import {
    AnalogMenuItem,
    BooleanMenuItem,
    EditableLargeNumberMenuItem, EditableTextMenuItem,
    EnumMenuItem, ListCreationMode,
    ListMenuItem,
    MenuItem, Rgb32MenuItem, ScrollChoiceMenuItem,
    SubMenuItem
} from "./MenuItem";
import {
    CodeGeneratorOptions,
    CreatorProperty,
    EepromSaveMode, MenuInMenuCollection,
    MenuInMenuConnectionType,
    MenuInMenuReplicationMode,
    MenuTreeWithCodeOptions,
    ProjectSaveLocation,
    RoundTripMode
} from "./ProjectStruct";
import {MenuTree} from "./MenuTree";
import {JsonMenuItemSerializer} from "./JsonMenuItemSerializer";

export const ANALOG_PERSIST_TYPE = "analogItem";
export const ENUM_PERSIST_TYPE = "enumItem";
export const SUB_PERSIST_TYPE = "subMenu";
export const ACTION_PERSIST_TYPE = "actionMenu";
export const RUNTIME_LIST_PERSIST_TYPE = "runtimeList";
export const CUSTOM_ITEM_PERSIST_TYPE = "customBuildItem";
export const BOOLEAN_PERSIST_TYPE = "boolItem";
export const TEXT_PERSIST_TYPE = "textItem";
export const FLOAT_PERSIST_TYPE = "floatItem";
export const RUNTIME_LARGE_NUM_PERSIST_TYPE = "largeNumItem";
export const SCROLL_CHOICE_PERSIST_TYPE = "scrollItem";
export const RGB32_COLOR_PERSIST_TYPE = "rgbItem";
export const TCMENU_COPY_PREFIX = "tcMenuCopy:";

export class PersistedMenu {
    parentId: string;
    readonly type: string;
    readonly item: MenuItem<any>;
    defaultValue?: string;

    constructor(parentId: string, type: string, item: MenuItem<any>, defaultValue?: string) {
        this.parentId = parentId;
        this.type = type;
        this.item = item;
        this.defaultValue = defaultValue;
    }
}

export function getPersistType(item: MenuItem<any>): string {
    // Some types might have different names in different parts of the system
    if (item.messageType === 'Analog') return ANALOG_PERSIST_TYPE;
    if (item.messageType === 'Enum') return ENUM_PERSIST_TYPE;
    if (item.messageType === 'Sub') return SUB_PERSIST_TYPE;
    if (item.messageType === 'Action') return ACTION_PERSIST_TYPE;
    if (item.messageType === 'List') return RUNTIME_LIST_PERSIST_TYPE;
    if (item.messageType === 'Boolean') return BOOLEAN_PERSIST_TYPE;
    if (item.messageType === 'Text') return TEXT_PERSIST_TYPE;
    if (item.messageType === 'Float') return FLOAT_PERSIST_TYPE;
    if (item.messageType === 'LargeNum') return RUNTIME_LARGE_NUM_PERSIST_TYPE;
    if (item.messageType === 'Scroll') return SCROLL_CHOICE_PERSIST_TYPE;
    if (item.messageType === 'Rgb32') return RGB32_COLOR_PERSIST_TYPE;
    if (item.messageType === 'CustomBuilder') return CUSTOM_ITEM_PERSIST_TYPE;

    return "unknown";
}

export interface PersistableStringList {
    id: string;
    listItems: string[];
}


export class PersistedProject {
    version: string;
    projectName: string;
    author: string;
    items: PersistedMenu[];
    codeOptions: CodeGeneratorOptions;
    stringLists: PersistableStringList[];

    constructor(version: string, projectName: string, author: string, items: PersistedMenu[],
                codeOptions: CodeGeneratorOptions, stringLists: PersistableStringList[]) {
        this.version = version;
        this.projectName = projectName;
        this.author = author;
        this.items = items;
        this.codeOptions = codeOptions;
        this.stringLists = stringLists;
    }
}

export function checkItemValueCanPersist(item: MenuItem<any>): boolean {
    return item instanceof AnalogMenuItem ||
        item instanceof EnumMenuItem ||
        item instanceof BooleanMenuItem ||
        item instanceof EditableLargeNumberMenuItem ||
        item instanceof Rgb32MenuItem ||
        item instanceof EditableTextMenuItem ||
        item instanceof ScrollChoiceMenuItem ;
}

function getProjectFromJson(json: any): PersistedProject {
    let serializer = new JsonMenuItemSerializer();
    let convertedItems: PersistedMenu[] = [];
    for(let item of json.items) {
        let menuItem = serializer.deserializeItem(item.type, item.item);
        if(menuItem) {
            convertedItems.push(new PersistedMenu(item.parentId.toString(), item.type, menuItem, item.defaultValue));
        }
    }

    const codeOptions = json.codeOptions as CodeGeneratorOptions;
    if (codeOptions) {
        if (typeof codeOptions.saveLocation === 'string') {
            codeOptions.saveLocation = ProjectSaveLocation[codeOptions.saveLocation as keyof typeof ProjectSaveLocation];
        }
        if(codeOptions.eepromSaveMode === undefined && codeOptions.usingSizedEEPROMStorage !== undefined) {
            codeOptions.eepromSaveMode = codeOptions.usingSizedEEPROMStorage ? EepromSaveMode.WRITE_BY_POSITION_WITH_SIZE : EepromSaveMode.LEGACY_WRITE_BY_POSITION;
        } else if (typeof codeOptions.eepromSaveMode === 'string') {
            codeOptions.eepromSaveMode = EepromSaveMode[codeOptions.eepromSaveMode as keyof typeof EepromSaveMode];
        }
        if (codeOptions.menuInMenuCollection && codeOptions.menuInMenuCollection.menuDefinitions) {
            for (let def of codeOptions.menuInMenuCollection.menuDefinitions) {
                if (typeof def.connectionType === 'string') {
                    def.connectionType = MenuInMenuConnectionType[def.connectionType as keyof typeof MenuInMenuConnectionType];
                }
                if (typeof def.replicationMode === 'string') {
                    def.replicationMode = MenuInMenuReplicationMode[def.replicationMode as keyof typeof MenuInMenuReplicationMode];
                }
            }
        }
    }

    return new PersistedProject(
        json.version, json.projectName, json.author, convertedItems,
        codeOptions, json.stringLists
    );
}

function applyCurrentValue(item: PersistedMenu, prj: PersistedProject) {
    if (item.item instanceof EnumMenuItem && item.defaultValue) {
        const idx = item.item.getItemList().indexOf(item.defaultValue);
        item.item.setCurrentValue(idx !== -1 ? idx : 0);
    } else {
        item.item.setCurrentValue(item.defaultValue);
    }

    if (item.item instanceof ListMenuItem && item.item.getListCreationMode() === ListCreationMode.FLASH_ARRAY && prj.stringLists) {
        const items = prj.stringLists
            .filter(psl => psl.id === item.item.getMenuId() && psl.listItems != null)
            .map(psl => psl.listItems)
            .at(0);
        if (items) {
            item.item.setCurrentValue(items);
        }
    }
}

export function parseEmfJsonToProject(jsonData: string, mode?: RoundTripMode) : MenuTreeWithCodeOptions {
    const json = JSON.parse(jsonData);
    const prj = getProjectFromJson(json);

    let tree = new MenuTree(menuTree => {});

    for(let item of prj.items) {
        tree.addMenuItem(item.parentId, item.item);
        if (item.defaultValue && checkItemValueCanPersist(item.item)) {
            applyCurrentValue(item, prj);
        }
    }
    return {
        menuTree: tree,
        description: prj.projectName,
        options: prj.codeOptions,
        roundTripMode: mode ?? RoundTripMode.EMF_FILE_DRAGGED_IN
    };

}

function copyOptionsToJsonSafe(options: CodeGeneratorOptions) {
    let optObj: any = {};
    Object.assign(optObj, options);
    if (options.saveLocation !== undefined) {
        optObj.saveLocation = ProjectSaveLocation[options.saveLocation];
    }
    if (options.eepromSaveMode !== undefined) {
        optObj.eepromSaveMode = EepromSaveMode[options.eepromSaveMode];
    }
    if (options.menuInMenuCollection && options.menuInMenuCollection.menuDefinitions) {
        optObj.menuInMenuCollection = {
            menuDefinitions: options.menuInMenuCollection.menuDefinitions.map(def => {
                let defObj: any = {};
                Object.assign(defObj, def);
                defObj.connectionType = MenuInMenuConnectionType[def.connectionType];
                defObj.replicationMode = MenuInMenuReplicationMode[def.replicationMode];
                return defObj;
            })
        };
    } else {
        optObj.menuInMenuCollection = {menuDefinitions: []} as MenuInMenuCollection;
    }
    return optObj;
}

function stringListsFromTree(allItems: MenuItem<any>[]) {
    const allLists = allItems.filter(item => item instanceof ListMenuItem && item.getListCreationMode() !== ListCreationMode.CUSTOM_RTCALL);
    return allLists.map(list => ({id: list.getMenuId(), listItems: list.getCurrentValue()}));
}

export function projectToPersistedJson(project: MenuTreeWithCodeOptions): any {
    const serializer = new JsonMenuItemSerializer();
    const items = serializer.populateListInOrder(project.menuTree.getRoot(), project.menuTree, false);
    return new PersistedProject(
        "1.00", // version
        project.description,
        "WebApp", // author
        items,
        copyOptionsToJsonSafe(project.options),
        stringListsFromTree(project.menuTree.getAllItems())
    );
}

File: api/tcMenuJavaApi/src/main/java/com/thecoderscorner/menu/persist/PersistedMenu.java
/*
 * Copyright (c)  2016-2019 https://www.thecoderscorner.com (Dave Cherry).
 * This product is licensed under an Apache license, see the LICENSE file in the top-level directory.
 *
 */

package com.thecoderscorner.menu.persist;

import com.thecoderscorner.menu.domain.*;
import com.thecoderscorner.menu.domain.util.MenuItemVisitor;

/**
 * Represents a persisted menu item, it has additional information needed to reconstitute the item at the right point
 * in the tree, namely the parentId, and also the type of menu item. This class is used by the JsonMenuItemSerializer
 * to store menu items.
 * @see JsonMenuItemSerializer
 */
public class PersistedMenu {
    public static final String ANALOG_PERSIST_TYPE = "analogItem";
    public static final String ENUM_PERSIST_TYPE = "enumItem";
    public static final String SUB_PERSIST_TYPE = "subMenu";
    public static final String ACTION_PERSIST_TYPE = "actionMenu";
    public static final String RUNTIME_LIST_PERSIST_TYPE = "runtimeList";
    public static final String CUSTOM_ITEM_PERSIST_TYPE = "customBuildItem";
    public static final String BOOLEAN_PERSIST_TYPE = "boolItem";
    public static final String TEXT_PERSIST_TYPE = "textItem";
    public static final String FLOAT_PERSIST_TYPE = "floatItem";
    public static final String RUNTIME_LARGE_NUM_PERSIST_TYPE = "largeNumItem";
    public static final String SCROLL_CHOICE_PERSIST_TYPE = "scrollItem";
    public static final String RGB32_COLOR_PERSIST_TYPE = "rgbItem";
    public static final String TCMENU_COPY_PREFIX = "tcMenuCopy:";

    private int parentId;
    private MenuItem item;
    private String type;
    private String defaultValue;

    public PersistedMenu() {
        // needed for Jackson
    }

    public PersistedMenu(MenuItem parent, MenuItem item) {
        this.parentId = parent.getId();
        this.item = item;
        item.accept(new MenuItemVisitor() {
            @Override
            public void visit(AnalogMenuItem item) {
                type = ANALOG_PERSIST_TYPE;
            }

            @Override
            public void visit(BooleanMenuItem item) {
                type = BOOLEAN_PERSIST_TYPE;
            }

            @Override
            public void visit(EnumMenuItem item) {
                type = ENUM_PERSIST_TYPE;
            }

            @Override
            public void visit(SubMenuItem item) {
                type = SUB_PERSIST_TYPE;
            }

            @Override
            public void visit(ActionMenuItem item) { type = ACTION_PERSIST_TYPE; }

            @Override
            public void visit(RuntimeListMenuItem listItem) {
                type = RUNTIME_LIST_PERSIST_TYPE;
            }

            @Override
            public void visit(CustomBuilderMenuItem listItem) {
                type = CUSTOM_ITEM_PERSIST_TYPE;
            }

            @Override
            public void visit(ScrollChoiceMenuItem scrollItem) {
                type = SCROLL_CHOICE_PERSIST_TYPE;
            }

            @Override
            public void visit(Rgb32MenuItem rgbItem) {
                type = RGB32_COLOR_PERSIST_TYPE;
            }

            @Override
            public void visit(EditableLargeNumberMenuItem numItem) {
                type = RUNTIME_LARGE_NUM_PERSIST_TYPE;
            }

            @Override
            public void visit(EditableTextMenuItem item) { type = TEXT_PERSIST_TYPE; }

            @Override
            public void visit(FloatMenuItem item) {
                type = FLOAT_PERSIST_TYPE;
            }
        });
    }

    public int getParentId() {
        return parentId;
    }

    public MenuItem getItem() {
        return item;
    }

    public void setParentId(int parentId) {
        this.parentId = parentId;
    }

    public void setItem(MenuItem item) {
        this.item = item;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getType() {
        return type;
    }

    public String getDefaultValue() {
        return defaultValue;
    }

    public void setDefaultValue(String def) {
        defaultValue = def;
    }
}
